### PR TITLE
[docs][expo-updates] Update spec for multipart manifest responses

### PR DESCRIPTION
**Pull request by wschurman**
# Why

This adds specification for the alternative `multipart/mixed` response structure for manifest responses, which is needed in order to send pieces of data to the client that shouldn't/can't be within the manifest itself since dynamic data within the manifest body makes end-to-end signing of the manifest body impossible.

Closes ENG-2558.

# How

Update spec.

# Test Plan

Read new spec, ensure it makes sense.

### CODE REVIEW

File: docs/pages/technical-specs/expo-updates-0.md [hunk 1]
```diff
@@ -44,25 +44,25 @@ A conformant client library MUST make a GET request with the headers:
...
+2. `expo-runtime-version` MUST be a runtime version compatible with the client. A runtime version stipulates the native code setup a client is running. It should be set when the client is built. For example, in an iOS client, the value may be set in a plist file.
 3. Any headers stipulated by a previous responses' [server defined headers](#manifest-response-headers).
 
-A conformant client library MUST also send at least one of `accept: application/expo+json` or `accept: application/json`, though SHOULD send `accept: application/expo+json, application/json` with the order following the preference ordering for the accept header specified in [RFC 7231](https://tools.ietf.org/html/rfc7231#section-5.3.2).
+A conformant client library MUST also send at least one of `accept: application/expo+json`, `accept: application/json`, or `accept: multipart/mixed` though SHOULD send `accept: application/expo+json, application/json,multipart/mixed` with the order following the preference ordering for the accept header specified in [RFC 7231](https://tools.ietf.org/html/rfc7231#section-5.3.2).
```
Comment (ide) [inline docs/pages/technical-specs/expo-updates-0.md:50]: The Accept header spec doesn't care about order. "q" relative weights matter most, follow by how specific the media type range is. If two media types have the same weight and specificity, they are equally preferred.
```suggestion
A conformant client library MUST also send at least one of `accept: application/expo+json`, `accept: application/json`, or `accept: multipart/mixed` though SHOULD send `accept: application/expo+json, application/json, multipart/mixed` with preferences expressed with "q" parameters as specified in [RFC 7231](https://tools.ietf.org/html/rfc7231#section-5.3.2).
```

File: docs/pages/technical-specs/expo-updates-0.md [hunk 2]
```diff
@@ -44,25 +44,25 @@ A conformant client library MUST make a GET request with the headers:
...
 expo-platform: *
 expo-runtime-version: *
 ```
 
 ## Manifest Response
 
-A conformant server MUST return a response containing a [manifest body](#manifest-response-body) and [headers](#manifest-response-headers).
+A conformant server MUST return a response structured in one of two ways:
+- For a response with `content-type: application/json` or `content-type: application/expo+json`, the [manifest headers](#manifest-response-headers) MUST be sent in the response headers and the [manifest body](#manifest-response-body) MUST be sent in the response body.
+- For a response with `content-type: multipart/mixed`, the response should be structured as specified in the [multipart manifest response](#multipart-manifest-response) section.
```
Comment (ide) [inline docs/pages/technical-specs/expo-updates-0.md:63]:
```suggestion
- For a response with `content-type: multipart/mixed`, the response MUST be structured as specified in the [multipart manifest response](#multipart-manifest-response) section.
```
?

File: docs/pages/technical-specs/expo-updates-0.md [hunk 3]
```diff
@@ -120,20 +113,50 @@ type Asset = {
...
     }
-    ...
   }
   ```
 
+### Multipart Manifest Response
+
+A manifest response of this format is defined by the `multipart/mixed` MIME type as defined by [RFC 2046](https://tools.ietf.org/html/rfc2046#section-5.1). Each part is defined as follows:
+1. REQUIRED `"manifest"` part:
+    - MUST have header `content-disposition: inline; name="manifest"` to identify this part.
```
Comment (ide) [inline docs/pages/technical-specs/expo-updates-0.md:133]: I wonder if we should loosen up the content-disposition value - partly to make it easier to use form data libraries

File: docs/pages/technical-specs/expo-updates-0.md [hunk 4]
```diff
@@ -120,20 +113,50 @@ type Asset = {
...
+
+Defined as JSON conforming to both the following `ManifestExtensions` definition expressed in [TypeScript](https://www.typescriptlang.org/) and the detailed descriptions for each field:
+```typescript
+type ManifestExtensions = {
+  assetRequestHeaders: ExpoAssetHeaderDictionary;
+  ...
+}
+
+type ExpoAssetHeaderDictionary = {
+  [key: <asset key>]: {
```
Comment (ide) [inline docs/pages/technical-specs/expo-updates-0.md:151]:
```suggestion
  [assetKey: string]: {
```

File: docs/pages/technical-specs/expo-updates-0.md [hunk 5]
```diff
@@ -143,11 +166,13 @@ accept: image/jpeg, */*
 accept-encoding: br, gzip
 ```
 
+A conformant client library SHOULD also include any header (key, value) pairs included in [`assetRequestHeaders`](#manifest-extensions) for this asset key.
```
Comment (ide) [inline docs/pages/technical-specs/expo-updates-0.md:169]: Should this say MUST instead of SHOULD?

**Review by jkhales — APPROVED** [review]

File: docs/pages/technical-specs/expo-updates-0.md [hunk 6]
```diff
@@ -44,25 +44,25 @@ A conformant client library MUST make a GET request with the headers:
...
 3. Any headers stipulated by a previous responses' [server defined headers](#manifest-response-headers).
 
-A conformant client library MUST also send at least one of `accept: application/expo+json` or `accept: application/json`, though SHOULD send `accept: application/expo+json, application/json` with the order following the preference ordering for the accept header specified in [RFC 7231](https://tools.ietf.org/html/rfc7231#section-5.3.2).
+A conformant client library MUST also send at least one of `accept: application/expo+json`, `accept: application/json`, or `accept: multipart/mixed` though SHOULD send `accept: application/expo+json, application/json,multipart/mixed` with the order following the preference ordering for the accept header specified in [RFC 7231](https://tools.ietf.org/html/rfc7231#section-5.3.2).
 
 Example:
 ```
-accept: application/expo+json, application/json
+accept: application/expo+json, application/json, multipart/mixed
```
[the remark this answers is not in the document]
Comment (wschurman) [reply docs/pages/technical-specs/expo-updates-0.md:54]: I don't think ordering matters in the header, and the only way to specify preferences is relative weights. I'll add these to the example.

File: docs/pages/technical-specs/expo-updates-0.md [hunk 7]
```diff
@@ -120,20 +113,50 @@ type Asset = {
...
   ```
 
+### Multipart Manifest Response
+
+A manifest response of this format is defined by the `multipart/mixed` MIME type as defined by [RFC 2046](https://tools.ietf.org/html/rfc2046#section-5.1). Each part is defined as follows:
+1. REQUIRED `"manifest"` part:
+    - MUST have header `content-disposition: inline; name="manifest"` to identify this part.
+    - The [manifest headers](#manifest-response-headers) MUST be sent in the part headers.
+    - The [manifest body](#manifest-response-body) MUST be sent in the part body.
+2. OPTIONAL `"extensions"` part:
```
[the remark this answers is not in the document]
Comment (jkhales) [reply docs/pages/technical-specs/expo-updates-0.md:136]: So the case you are thinking about is the client asks for multipart, but unsigned? I don't understand this use case🤔

> conformant servers to only implement the mixed response type
as its written, this doesn't seem allowed by the spec?

File: docs/pages/technical-specs/expo-updates-0.md [hunk 8]
```diff
@@ -44,25 +44,25 @@ A conformant client library MUST make a GET request with the headers:
...
 3. Any headers stipulated by a previous responses' [server defined headers](#manifest-response-headers).
 
-A conformant client library MUST also send at least one of `accept: application/expo+json` or `accept: application/json`, though SHOULD send `accept: application/expo+json, application/json` with the order following the preference ordering for the accept header specified in [RFC 7231](https://tools.ietf.org/html/rfc7231#section-5.3.2).
+A conformant client library MUST also send at least one of `accept: application/expo+json`, `accept: application/json`, or `accept: multipart/mixed` though SHOULD send `accept: application/expo+json, application/json, multipart/mixed` with preferences expressed with "q" parameters as specified in [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.1).
 
 Example:
 ```
-accept: application/expo+json, application/json
+accept: application/expo+json;q=0.9, application/json;q=0.8, multipart/mixed
```
Comment (ide) [inline docs/pages/technical-specs/expo-updates-0.md:54]: Might be clearer to put multipart/mixed first or add a q value. q=1.0 is the default if unspecified. Or just leave off all the q values since they don't matter here.

File: docs/pages/technical-specs/expo-updates-0.md [hunk 9]
```diff
@@ -120,20 +113,50 @@ type Asset = {
...
 
+### Multipart Manifest Response
+
+A manifest response of this format is defined by the `multipart/mixed` MIME type as defined by [RFC 2046](https://tools.ietf.org/html/rfc2046#section-5.1). Each part is defined as follows:
+1. REQUIRED `"manifest"` part:
+    - MUST have a `content-disposition` header with a name parameter equal to `manifest`. e.g. `content-disposition: inline; name="manifest"`
+    - The [manifest headers](#manifest-response-headers) MUST be sent in the part headers.
+    - The [manifest body](#manifest-response-body) MUST be sent in the part body.
+2. OPTIONAL `"extensions"` part:
+    - MUST have `content-disposition` header with a name parameter equal to `extensions`. e.g. `content-disposition: inline; name="extensions"`
```
Comment (ide) [inline docs/pages/technical-specs/expo-updates-0.md:137]:
```suggestion
    - MUST have `content-disposition` header with a name parameter equal to `extensions`. E.g. `content-disposition: inline; name="extensions"`.
```

File: docs/pages/technical-specs/expo-updates-0.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (wschurman) [reply docs/pages/technical-specs/expo-updates-0.md:136]: You bring up a good point which is the question: is it considered conformant if a client and/or server implementation don't support multipart or don't support normal JSON? Is sending a 406 in either of these cases conformant?

The next logical question is whether we want our implementations to be conformant (we should IMO). Our client implementation (the expo client) will support both and `accept` both. We were thinking of making our server only support multipart since we need it for asset HMACs and all our assets need HMACs (we can't serve a normal manifest). But then comes a question: what if a custom client implementation wants to talk to the EAS server implementation? How do we specify in the spec that this is possible even if we can't serve normal JSON?

[the remark this answers is not in the document]
Comment (ide) [reply docs/pages/technical-specs/expo-updates-0.md:136]: We probably will change the spec to say the server can respond with a 406 or the content type it supports.

An extensionless multipart response doesn't cause any problems and doesn't make the client code more difficult to write since we need to handle the case where "extensions" could be missing (don't want to crash the app).

[the remark this answers is not in the document]
Comment (wschurman) [reply docs/pages/technical-specs/expo-updates-0.md:136]: > We probably will change the spec to say the server can respond with a 406 or the content type it supports.

This spec change is probably the most appropriate time and place to make this decision.

- If we say that a conformant server can respond with a 406, that means that it is ok for a conformant client and conformant server never to be able to communicate and get updates which IMO seems to be a pretty big issue.
- If we say that a conformant server can respond with a content type it supports even if it wasn't in the request accept header, then that means it is ok for a conformant client to receive an update it doesn't know how to comprehend.
- Another option is that we could say that a conformant client must support both JSON and multipart and always `accept` both.

File: docs/pages/technical-specs/expo-updates-0.md (hunk 8 above)
Comment (wschurman) [reply docs/pages/technical-specs/expo-updates-0.md:54]: I actually think this example is useful to show that:
a) order doesn't matter
b) it defaults to 1.0

i.e. it's useful to show that the other spec is being followed.

File: docs/pages/technical-specs/expo-updates-0.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ide) [reply docs/pages/technical-specs/expo-updates-0.md:136]: I don't see the 406 as a big issue. You'd find this out in development when picking what server to use.

Clients already need to handle responses they don't understand e.g. the 4xx HTML pages from NGINX.

[the remark this answers is not in the document]
Comment (wschurman) [reply docs/pages/technical-specs/expo-updates-0.md:136]: Yeah, that's fair. I'll add a blurb along the lines of "A conformant server may support either or both, and a client may support either or both. When a server is unable to respond to a client, a 406 should be returned".

File: docs/pages/technical-specs/expo-updates-0.md (hunk 8 above)
Comment (ide) [reply docs/pages/technical-specs/expo-updates-0.md:54]: OK, in that case I think it is useful to say in plain English what the preference ordering is.

File: docs/pages/technical-specs/expo-updates-0.md [hunk 10]
```diff
@@ -44,25 +44,25 @@ A conformant client library MUST make a GET request with the headers:
...
-accept: application/expo+json, application/json
+accept: application/expo+json;q=0.9, application/json;q=0.8, multipart/mixed
 expo-platform: *
 expo-runtime-version: *
 ```
 
 ## Manifest Response
 
-A conformant server MUST return a response containing a [manifest body](#manifest-response-body) and [headers](#manifest-response-headers).
+A conformant server MUST return a response structured in at least one of two ways. A conformant server MAY support either or both response structures, and when an unsupported response structure is requested the server should respond with a `406` error.
```
Comment (ide) [inline docs/pages/technical-specs/expo-updates-0.md:61]:
```suggestion
A conformant server MUST return a response structured in at least one of two ways. A conformant server MAY support either or both response structures, and when an unsupported response structure is requested the server SHOULD respond with an HTTP `406` error status.
```

**Review by ide — APPROVED** [review]

File: docs/pages/technical-specs/expo-updates-0.md [hunk 11]
```diff
@@ -44,25 +44,25 @@ A conformant client library MUST make a GET request with the headers:
...
+2. `expo-runtime-version` MUST be a runtime version compatible with the client. A runtime version stipulates the native code setup a client is running. It should be set when the client is built. For example, in an iOS client, the value may be set in a plist file.
 3. Any headers stipulated by a previous responses' [server defined headers](#manifest-response-headers).
 
-A conformant client library MUST also send at least one of `accept: application/expo+json` or `accept: application/json`, though SHOULD send `accept: application/expo+json, application/json` with the order following the preference ordering for the accept header specified in [RFC 7231](https://tools.ietf.org/html/rfc7231#section-5.3.2).
+A conformant client library MUST also send at least one of `accept: application/expo+json`, `accept: application/json`, or `accept: multipart/mixed` based on the supported response structures though SHOULD send `accept: application/expo+json, application/json, multipart/mixed`. A conformant client library MAY express preference using "q" parameters as specified in [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.1), which default to `1`.
```
Comment (quinlanj) [inline docs/pages/technical-specs/expo-updates-0.md:50]: `application/json` will only work for classic updates and the self hosted servers. I guess this comes with the caveat in the later section, that servers can respond with 406 if they don't support the format.

**Review by quinlanj — APPROVED** [review]
